Implement Middleware for Handlers

gkinsman opened the pull request:
#### This PR implements a middleware layer around handlers. Middlewares have a number of advantages over existing solutions:

- User provided middlewares can be provided via extension methods packages in nuget packages
- They are composable with other middlewares
- They don't require any custom code in or around `IHandlerAsync<T>`
- They are cancellable with `CancellationToken`, whereas `IHandlerAsync<T>` cannot change to add one (or at least we haven't decided to do that)
- They create a nice place for users to add in cross cutting code, like custom stats, retry strategies, locking, etc etc without needing to change JustSaying.

This work is largely based on the existing middleware code added by @maisiesadler that is used in the `MessageReceiveBuffer`'s receive pipeline.

### Some example code

[Middlewares can be pulled from DI, or passed in explicitly](https://github.com/gkinsman/JustSaying/blob/31502b59ff0544c22d04ae6e1dead7f3a8c8e717/tests/JustSaying.IntegrationTests/Fluent/Subscribing/WhenUsingMultipleMiddlewares.cs#L35-L41):
```csharp
var callRecord = new List<string>();

void Before(string id) => callRecord.Add($"Before_{id}");
void After(string id) => callRecord.Add($"After_{id}");

var outerMiddleware = new TrackingMiddleware("outer", Before, After);
var middleMiddleware = new TrackingMiddleware("middle", Before, After);
var innerMiddleware = new TrackingMiddleware("inner", Before, After);

var services = GivenJustSaying()
    .AddSingleton(outerMiddleware)
    .AddSingleton<IHandlerAsync<SimpleMessage>>(handler)
    .ConfigureJustSaying(builder => builder.WithLoopbackTopic<SimpleMessage>(UniqueName,
        topic => topic.WithMiddlewareConfiguration(app =>
        {
            app.Use<TrackingMiddleware>(); // from DI
            app.Use(() => middleMiddleware); // provide a Func<MiddlewareBase<HandleMessageContext, bool>
            app.Use(innerMiddleware); // Existing instance
        })));
```

Produces:
```
Before_outer
Before_middle
Before_inner
After_inner
After_middle
After_outer
```

[Implementation of an extension method that makes a `UseExactlyOnce` method available to the pipeline, and resolves some services from DI](https://github.com/gkinsman/JustSaying/blob/770a129bbb590be5a795c514cf63235d29f237d7/src/JustSaying/Messaging/Middleware/Handle/HandlerMiddlewareBuilderExtensions.cs#L23-L41)

```csharp
public static HandlerMiddlewareBuilder UseExactlyOnce<TMessage>(
    this HandlerMiddlewareBuilder builder,
    string lockKey,
    TimeSpan? lockDuration = null)
{
    HandleMessageMiddleware CreateMiddleware()
    {
        var messageLock = builder.ServicesBuilder?.MessageLock?.Invoke() ?? builder.ServiceResolver.ResolveService<IMessageLockAsync>();
        var logger = builder.ServiceResolver.ResolveService<ILogger<ExactlyOnceMiddleware<TMessage>>>();

        return new ExactlyOnceMiddleware<TMessage>(messageLock,
            lockDuration ?? TimeSpan.MaxValue,
            lockKey,
            logger);
    }

    builder.Use(CreateMiddleware);

    return builder;
}
```

[Usage of `UseExactlyOnce`](https://github.com/gkinsman/JustSaying/blob/3f71f09917074881e678e573bf1cedeeb73130e8/tests/JustSaying.IntegrationTests/Fluent/Subscribing/WhenHandlerIsDeclaredAsExactlyOnce.cs#L29-L33)

```csharp
.ConfigureJustSaying((builder) =>
    builder.WithLoopbackTopic<SimpleMessage>(UniqueName,
        c =>
            c.WithMiddlewareConfiguration(m =>
                m.UseExactlyOnce<SimpleMessage>("simple-message-lock"))))
```

[They can be chained](https://github.com/gkinsman/JustSaying/blob/770a129bbb590be5a795c514cf63235d29f237d7/tests/JustSaying.UnitTests/Messaging/Middleware/HandlerMiddlewareBuilderTests.cs#L34-L56):
```csharp
.Configure(pipe =>
                {
                    pipe.Use(m1);
                    pipe.Use(m2);
                    pipe.Use(m3);
                })
```

slang25 commented on src/JustSaying/Messaging/Middleware/Handle/HandlerInvocationMiddleware.cs:
```diff
@@ -0,0 +1,44 @@
+            IHandlerAsync<T> handler = _handlerResolver(resolutionContext);
+
+            //TODO: cache these
+            Type handlerType = handler.GetType();
```
Can we instead use `var task = handler.Handle(context.MessageAs<T>());` and flow the generic constraint?

gkinsman replied:
Ahh, yes we can! I did it that way initially but then wanted to remove the generic parameter from `HandlerInvocationMiddleware<T>` for some reason, but that turned out to not be necessary.

Great stuff!

gkinsman replied:
Have fixed this up 👍

gkinsman commented:
@martincostello can we target this for post-beta? I'm really keen to get the beta process rolling so we can get some features using the new backend.

martincostello commented on src/JustSaying/Messaging/Middleware/Handle/HandlerMiddlewareBuilderExtensions.cs:
```diff
@@ -0,0 +1,45 @@
+using JustSaying.Models;
+using Microsoft.Extensions.Logging;
+
+using HandleMessageMiddleware = JustSaying.Messaging.Middleware.MiddlewareBase<JustSaying.Messaging.Middleware.Handle.HandleMessageContext, bool>;
+
+
```
```suggestion
using HandleMessageMiddleware = JustSaying.Messaging.Middleware.MiddlewareBase<JustSaying.Messaging.Middleware.Handle.HandleMessageContext, bool>;

```

martincostello commented on src/JustSaying/Messaging/Middleware/Handle/HandlerMiddlewareBuilderExtensions.cs:
```diff
@@ -0,0 +1,45 @@
+        public static HandlerMiddlewareBuilder UseHandler<TMessage>(this HandlerMiddlewareBuilder builder,
+            Func<HandlerResolutionContext, IHandlerAsync<TMessage>> handler) where TMessage : Message
+        {
+            if(builder == null) throw new ArgumentNullException(nameof(builder));
+            if(handler == null) throw new ArgumentNullException(nameof(handler));
```
```suggestion
            if (builder == null) throw new ArgumentNullException(nameof(builder));
            if (handler == null) throw new ArgumentNullException(nameof(handler));
```

martincostello commented:
> @martincostello can we target this for post-beta? I'm really keen to get the beta process rolling so we can get some features using the new backend.

We can have as many alpha/beta/gamma/whatevers as we want before release internally, and we have the ProGet feed we can host packages in, so I don't think it really matters what label we give it or whatever the release cadence is (or the need to be super-strict on SemVer).

I'd only worry about these distinctions for putting something in NuGet.org, and I think we'd want to be fairly confidently "done" and happy it's working before publishing a beta there.

gkinsman commented:
> > @martincostello can we target this for post-beta? I'm really keen to get the beta process rolling so we can get some features using the new backend.
> 
> We can have as many alpha/beta/gamma/whatevers as we want before release internally, and we have the ProGet feed we can host packages in, so I don't think it really matters what label we give it or whatever the release cadence is (or the need to be super-strict on SemVer).
> 
> I'd only worry about these distinctions for putting something in NuGet.org, and I think we'd want to be fairly confidently "done" and happy it's working before publishing a beta there.

That makes perfect sense, thanks 👍. The main reason for a beta was so that we can switch to a package feed instead of local packages, so as you say we can just use proget 🙂.

gkinsman commented on src/JustSaying/Fluent/ISubscriptionBuilder`1.cs:
```diff
@@ -26,7 +26,9 @@ internal interface ISubscriptionBuilder<out T>
         void Configure(
             JustSayingBus bus,
             IHandlerResolver resolver,
+            IServiceResolver serviceResolver,
```
Yeah I agree that having both resolvers plus the ServicesBuilder here isn't ideal. We could perhaps introduce an `IMiddlewareResolver`, though that adds a bit more machinery for questionable gain.

The ServicesBuilder is needed because we special case the registration/resolution of `IMessageLockAsync` because of JustSaying's explicit support for ExactlyOnce. We could remove this dependency by removing the `WithMessageLock` API and just requiring the user to register an `IMessageLockAsync` into the DI container, which would be resolvable in the middleware.

I suspect we might need to keep `IHandlerResolver` and `IServiceResolver` though, as a custom `IHandlerResolver` could be provided even if the user uses a DI container - we can't just resolve the `IHandlerResolver` from the `IServiceResolver`

gkinsman replied:
I had a punt at doing this over here https://github.com/gkinsman/JustSaying/pull/1

gkinsman replied:
Considering this resolved 👍

gkinsman commented on src/JustSaying/Messaging/Middleware/Handle/HandleMessageContext.cs:
```diff
@@ -0,0 +1,23 @@
+
+        public Message Message { get; }
+
+        public TMessage MessageAs<TMessage>() where TMessage : Message => Message as TMessage;
```
Have moved this out to an extension method. I think it's a useful hint to middleware authors that it's possible to pull out a Message of type `T` (which is available on the middleware).

adammorr commented on src/JustSaying/AwsTools/MessageHandling/Dispatch/MessageDispatcher.cs:
```diff
@@ -157,13 +159,13 @@ public class MessageDispatcher : IMessageDispatcher
+            if (middleware == null)
             {
                 _logger.LogError(
                     "Failed to dispatch. Handler for message of type '{MessageTypeName}' not found in handler map.",
```
Do you think this log message needs updating now that we're using middleware instead?

adammorr replied:
Just re-read your description, I'll hold off on these things until it's ready for review

gkinsman commented:
This is now ready for review! I've documented the public interfaces, added tests, and removed the old exactly-once and stopwatch handlers and reimplemented as middlewares.

gkinsman commented on tests/JustSaying.TestingFramework/JustSaying.TestingFramework.csproj:
```diff
@@ -11,7 +11,7 @@
     <PackageReference Include="Microsoft.Extensions.Logging" Version="3.1.8" />
     <PackageReference Include="NSubstitute" Version="4.2.2" />
-    <PackageReference Include="Shouldly" Version="4.0.0" />
+    <PackageReference Include="Shouldly" Version="3.0.2" />
```
I downgraded this because v4 uses a new approval test engine called DiffEngine that triggers a type load exception for some reason, in a dependent lib called EmptyFiles (https://www.nuget.org/packages/EmptyFiles/). I couldn't get to the bottom of it so downgraded for now.

gkinsman replied:
Ah, this looks like an issue with `net5.0` and Shouldly 4.0 - it works in `netcoreapp3.1`. Will keep this downgraded till there's a fix.

slang25 replied:
I'm trying to recreate this so I can file a bug report, but haven't been able to yet, is there a commit in this PR I should use?

gkinsman replied:
Hey 🙂, I managed to reproduce it before over here: https://github.com/gkinsman/ShouldlyRepro. It fails locally with `dotnet test` and in Rider, but succeeds when run in github actions - weird!

I was going to log a bug but hadn't quite narrowed down if it was my environment yet. Does that project build locally for you?

slang25 replied:
Thanks @gkinsman, that was really helpful: https://github.com/shouldly/shouldly/issues/717

slang25 replied:
We can bump this back to `4.0.1` now

slang25 commented on src/JustSaying/JustSayingBus.cs:
```diff
@@ -46,15 +49,9 @@ public IMessageMonitor Monitor
         public IMessageContextAccessor MessageContextAccessor { get; set; }
 
-        public HandlerMap HandlerMap { get; }
+        public MiddlewareMap MiddlewareMap { get; }
```
If we hid this from the public API we could make `MiddlewareMap` internal (we'd need to also make `MessageDispatcher` internal, which is probably a good idea).
Perhaps we evaluate what we can hide as a separate task outside of this PR.

gkinsman replied:
We can make this internal while leaving the MessageDispatcher as is - MiddlewareMap doesn't have the same strange dependencies that HandlerMap had because it doesn't explicitly know about message locks anymore.

I think its still a good idea to review visibility separately - we might be able to hide some more things now that we couldn't before 👍

slang25 commented on src/JustSaying/Messaging/Middleware/Handle/HandlerMiddlewareBuilder.cs:
```diff
@@ -0,0 +1,139 @@
+        /// <exception cref="InvalidOperationException">
+        /// If a HandlerInvocationMiddleware already exists in this pipeline, it cannot be added again.
+        /// </exception>
+        public HandlerMiddlewareBuilder UseHandler<TMessage>() where TMessage : Message
```
`internal` perhaps?

gkinsman replied:
This has been changed to internal. If in the future we want to open up the whole pipeline for customisation (so e.g. you could use a delegate as a handler instead of an `IHandlerAsync`) then we can re-think how we supply this pipeline through the API, but this should get us started.

slang25 commented on src/JustSaying/Messaging/Middleware/Metrics/StopwatchMiddleware.cs:
```diff
@@ -0,0 +1,43 @@
+namespace JustSaying.Messaging.Middleware.Metrics
+{
+    /// <summary>
+    /// This middleware measures handlers' execution duration and reports the results to <see cref="IMessageMonitor"/>.
```
```suggestion
    /// This middleware measures the handler's execution duration and reports the results to <see cref="IMessageMonitor"/>.
```
(this is measuring 1 handler, not multiple right?)

gkinsman replied:
Yep, think yours is clearer 👍

slang25 reviewed (commented):
Some minor comments, overall a really good addition 😃

martincostello commented on src/JustSaying/AwsTools/MessageHandling/Dispatch/MiddlewareMap.cs:
```diff
@@ -0,0 +1,103 @@
+
+        public InterrogationResult Interrogate()
+        {
+            var middlewares = _middlewares.Select(item =>
```
You should `ToList()` this, otherwise it's not materialized when it's returned and could cause different results if called multiple times and/or `InvalidOperationException` to be thrown if an item is added to the dictionary mid-enumeration.

gkinsman replied:
Ah good catch, thanks 👍

martincostello commented on src/JustSaying/Fluent/SqsReadConfigurationBuilder.cs:
```diff
@@ -20,12 +21,53 @@ public sealed class SqsReadConfigurationBuilder : SqsConfigurationBuilder<SqsRea
+        /// });
+        /// </code>
+        /// would yield this order of execution:
+        /// Before_SomeCustomMiddleware
```
Could format this as an HTML `<ul>`?

gkinsman replied:
We could do - this is the appearance in rider(half with, half without):

![image](https://user-images.githubusercontent.com/130231/99518181-25b76880-2988-11eb-9b4d-bb64185d24d7.png)

I checked the formatting in VS and Rider, and had to change a few names to make it format correcly, but maybe a ul is a better way to go.

gkinsman replied:
Have moved this to ul for consistent formatting across ide's

martincostello commented on src/JustSaying/Fluent/SqsReadConfigurationBuilder.cs:
```diff
@@ -20,12 +21,53 @@ public sealed class SqsReadConfigurationBuilder : SqsConfigurationBuilder<SqsRea
+        /// After_SomeOtherCustomMiddleware
+        /// After_SomeCustomMiddleware
+        /// </example>
+        /// <returns></returns>
```
Missing returns value docs.

martincostello commented on src/JustSaying/JustSaying.csproj:
```diff
@@ -14,4 +14,7 @@
     <PackageReference Include="System.Text.Json" Version="4.6.0" />
     <PackageReference Include="System.Threading.Channels" Version="4.5.0" />
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Messaging\Middleware\Interrogation" />
+  </ItemGroup>
```
```suggestion
```

martincostello commented on src/JustSaying/Messaging/Middleware/ExactlyOnce/ExactlyOnceMiddleware.cs:
```diff
@@ -1,38 +1,38 @@
 using Microsoft.Extensions.Logging;
 
-namespace JustSaying.Messaging.MessageHandling
+namespace JustSaying.Messaging.Middleware.ExactlyOnce
```
Does this need to live in its own namespace? Seems a bit OTT for one class and one extension method to me.

gkinsman replied:
You might be right - I was thinking that it'd be good to keep these separate as we add more (e.g. a polly retry/rate limiting), but having separate namespaces is maybe a bit unnecessary.

I've moved everything into a base `JustSaying.Messaging.Middleware` namespace while retaining the folder structure to keep things tidy, which makes things a bit more discoverable as you noted below.

What do you think?

martincostello commented on src/JustSaying/Messaging/Middleware/ExactlyOnce/HandlerMiddlewareBuilderExtensions.cs:
```diff
@@ -0,0 +1,40 @@
+            this HandlerMiddlewareBuilder builder,
+            string lockKey,
+            TimeSpan? lockDuration = null)
+        {
```
Null check the builder (and the key?).

martincostello commented on src/JustSaying/Messaging/Middleware/ExactlyOnce/HandlerMiddlewareBuilderExtensions.cs:
```diff
@@ -0,0 +1,40 @@
+using Microsoft.Extensions.Logging;
+using HandleMessageMiddleware = JustSaying.Messaging.Middleware.MiddlewareBase<JustSaying.Messaging.Middleware.Handle.HandleMessageContext, bool>;
+
+namespace JustSaying.Messaging.Middleware.ExactlyOnce
```
This API would be more discoverable if it was in the same namespace as the builder it's extending.

gkinsman replied:
Done 👍

martincostello commented on src/JustSaying/Messaging/Middleware/Handle/HandleMessageContext.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace JustSaying.Messaging.Middleware.Handle
+{
+    public sealed class HandleMessageContext
```
`///` docs

martincostello commented on src/JustSaying/Messaging/Middleware/Handle/HandleMessageContextExtensions.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace JustSaying.Messaging.Middleware.Handle
+{
+    public static class HandleMessageContextExtensions
```
`///` docs

martincostello commented on src/JustSaying/Messaging/Middleware/Handle/HandleMessageContextExtensions.cs:
```diff
@@ -0,0 +1,16 @@
+
```
```suggestion
```

martincostello commented on src/JustSaying/Messaging/Middleware/Handle/HandleMessageContextExtensions.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        public static TMessage MessageAs<TMessage>(this HandleMessageContext context) where TMessage : Message
+        {
+            if(context == null) throw new ArgumentNullException(nameof(context));
```
```suggestion
            if (context == null) throw new ArgumentNullException(nameof(context));
```

martincostello commented on src/JustSaying/Messaging/Middleware/Handle/HandlerMiddlewareBuilder.cs:
```diff
@@ -0,0 +1,139 @@
+        {
+            if (_handlerMiddleware != null)
+            {
+                throw new InvalidOperationException($"Handler middleware has already been specified " +
```
Do we need to split this message over two lines?

gkinsman replied:
Moved it to a single line 👍

martincostello commented on src/JustSaying/Messaging/Middleware/Handle/HandlerMiddlewareBuilderExtensions.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace JustSaying.Messaging.Middleware.Handle
+{
+
```
```suggestion
```

martincostello commented on src/JustSaying/Messaging/Middleware/Handle/HandlerMiddlewareBuilderExtensions.cs:
```diff
@@ -0,0 +1,36 @@
+        /// <typeparam name="TMessage">The type of the message that should be handled</typeparam>
+        /// <returns>The current <see cref="HandlerMiddlewareBuilder"/>.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="builder"/> is <see langword="null"/>.
```
```suggestion
        /// <paramref name="builder"/> or <paramref name="handler"/> is <see langword="null"/>.
```

martincostello commented on src/JustSaying/Messaging/Middleware/Metrics/HandlerMiddlewareBuilderExtensions.cs:
```diff
@@ -0,0 +1,29 @@
+using JustSaying.Messaging.Middleware.Handle;
+using JustSaying.Messaging.Monitoring;
+
+namespace JustSaying.Messaging.Middleware.Metrics
```
Similar comments about discoverability putting everything in discrete namespaces.

gkinsman replied:
These are all now in the same namespace, `JustSaying.Messaging.Middleware`, and I've marked the folders as not contributing to namespaces to avoid the squiggly in jetbrains products.

martincostello commented on src/JustSaying/Messaging/Middleware/Metrics/StopwatchMiddleware.cs:
```diff
@@ -0,0 +1,43 @@
+
+            bool result = await func(stoppingToken).ConfigureAwait(false);
+
+            watch.Stop();
```
The timing won't be recorded if the func throws, is that intentional?

gkinsman replied:
So this is the same logic as the previous handler, [found here](https://github.com/justeat/JustSaying/blob/master/src/JustSaying/Messaging/Monitoring/StopwatchHandler.cs).

I guess it makes sense if you assume that `IMessageMonitor` only calls `HandlerExecutionTime` for successful executions. I'm not sure if that's what I'd expect, but its the way it works now.

Should we change this?

martincostello replied:
If it's consistent with existing behaviour, then that's fine.

martincostello commented on tests/JustSaying.UnitTests/JustSaying.UnitTests.csproj:
```diff
@@ -19,4 +19,9 @@
     <PackageReference Include="xunit" Version="2.4.1" />
     <PackageReference Include="xunit.runner.visualstudio" Version="2.4.3" />
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Messaging\Channels\SubscriptionGroupTests\Support" />
+    <Folder Include="Messaging\Middleware\Approvals" />
+    <Folder Include="Messaging\Monitoring" />
+  </ItemGroup>
```
Are these needed?

gkinsman replied:
Nope, will remove 👍

martincostello commented on src/JustSaying/JustSaying.csproj.DotSettings:
```diff
@@ -0,0 +1,4 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
Maybe this should be in the `.gitignore`?

gkinsman replied:
It looks like this file is intended to be for user settings - I thought it was a project settings file. I've removed it and moved the rule exclusions into code using `// ReSharper disable once CheckNamespace` to avoid the squiggly.

martincostello commented on src/JustSaying/Messaging/Middleware/ExactlyOnce/ExactlyOnceHandlerMiddlewareBuilderExtensions.cs:
```diff
@@ -21,6 +20,9 @@ public static class HandlerMiddlewareBuilderExtensions
             string lockKey,
             TimeSpan? lockDuration = null)
         {
+            if(builder == null) throw new ArgumentNullException(nameof(builder));
+            if(string.IsNullOrEmpty(lockKey)) throw new ArgumentException("Parameter cannot be null or empty", nameof(lockKey));
```
```suggestion
            if (builder == null) throw new ArgumentNullException(nameof(builder));
            if (string.IsNullOrEmpty(lockKey)) throw new ArgumentException("Parameter cannot be null or empty.", nameof(lockKey));
```
